diff --git a/include/mbgl/style/sources/tile_source.hpp b/include/mbgl/style/sources/tile_source.hpp
index 3516ee7..fcb91f7 100644
--- a/include/mbgl/style/sources/tile_source.hpp
+++ b/include/mbgl/style/sources/tile_source.hpp
@@ -17,6 +17,7 @@ public:
 
     const variant<std::string, Tileset>& getURLOrTileset() const;
     std::optional<std::string> getURL() const;
+    const std::optional<Tileset>& getTileset() const;
 
     uint16_t getTileSize() const;
 
diff --git a/src/mbgl/style/sources/tile_source.cpp b/src/mbgl/style/sources/tile_source.cpp
index 5eb1872..4c6859c 100644
--- a/src/mbgl/style/sources/tile_source.cpp
+++ b/src/mbgl/style/sources/tile_source.cpp
@@ -38,6 +38,10 @@ std::optional<std::string> TileSource::getURL() const {
     return urlOrTileset.get<std::string>();
 }
 
+const std::optional<Tileset>& TileSource::getTileset() const {
+    return impl().tileset;
+}
+
 uint16_t TileSource::getTileSize() const {
     return impl().getTileSize();
 }
