diff --git a/include/mln/shaders/webgpu/location_indicator.hpp b/include/mln/shaders/webgpu/location_indicator.hpp
index 14554a162da5f8cba2596e63f9d69a82d3844a7d..2cc9d4617a74c71f650f8ab6b4388b67084e3db2 100644
--- a/include/mln/shaders/webgpu/location_indicator.hpp
+++ b/include/mln/shaders/webgpu/location_indicator.hpp
@@ -29,5 +29,5 @@ struct LocationIndicatorUBO {
 };
 
-@group(0) @binding(0) var<uniform> ubo: LocationIndicatorUBO;
+@group(0) @binding(4) var<uniform> ubo: LocationIndicatorUBO;
 
 @vertex
@@ -45,5 +45,5 @@ struct LocationIndicatorUBO {
 };
 
-@group(0) @binding(0) var<uniform> ubo: LocationIndicatorUBO;
+@group(0) @binding(4) var<uniform> ubo: LocationIndicatorUBO;
 
 @fragment
@@ -77,5 +77,5 @@ struct LocationIndicatorUBO {
 };
 
-@group(0) @binding(0) var<uniform> ubo: LocationIndicatorUBO;
+@group(0) @binding(4) var<uniform> ubo: LocationIndicatorUBO;
 
 @vertex
@@ -94,5 +94,5 @@ struct LocationIndicatorUBO {
 };
 
-@group(0) @binding(0) var<uniform> ubo: LocationIndicatorUBO;
+@group(0) @binding(4) var<uniform> ubo: LocationIndicatorUBO;
 @group(0) @binding(1) var colorTexture: texture_2d<f32>;
 @group(0) @binding(2) var colorSampler: sampler;
