diff --git a/CMakeLists.txt b/CMakeLists.txt
index 77756c64ef19..4e3ba384640a 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -2,6 +2,7 @@ cmake_minimum_required(VERSION 3.25 FATAL_ERROR)
 
 option(MLN_WITH_CORE_ONLY "Build only the core bits, no platform code" OFF)
 option(MLN_WITH_PLUGINS "Build source-bound native plugin layers" OFF)
+option(MLN_WITH_LEGACY_ANNOTATIONS "Build the legacy annotation API and add its source and layer to every style" ON)
 option(MLN_WITH_CLANG_TIDY "Build with clang-tidy checks enabled" OFF)
 option(MLN_WITH_COVERAGE "Enable code coverage collection" OFF)
 option(MLN_WITH_QT "Build MapLibre Native Qt bindings" OFF)
@@ -156,6 +157,13 @@ endif()
 
 include(cmake/plugins.cmake)
 
+if(NOT MLN_WITH_LEGACY_ANNOTATIONS)
+    target_compile_definitions(
+        mbgl-core
+        PUBLIC MLN_LEGACY_ANNOTATIONS_DISABLED
+    )
+endif()
+
 # Override visibility for mbgl-core when building amalgamation
 if(MLN_CREATE_AMALGAMATION)
     target_compile_options(mbgl-core PRIVATE
diff --git a/platform/default/src/mln/layermanager/layer_manager.cpp b/platform/default/src/mln/layermanager/layer_manager.cpp
index bfad3dfcd8c6..891b56de761e 100644
--- a/platform/default/src/mln/layermanager/layer_manager.cpp
+++ b/platform/default/src/mln/layermanager/layer_manager.cpp
@@ -123,7 +123,7 @@ LayerManager* LayerManager::get() noexcept {
 }
 
 #if defined(MBGL_LAYER_LINE_DISABLE_ALL) || defined(MBGL_LAYER_SYMBOL_DISABLE_ALL) || \
-    defined(MBGL_LAYER_FILL_DISABLE_ALL)
+    defined(MBGL_LAYER_FILL_DISABLE_ALL) || defined(MLN_LEGACY_ANNOTATIONS_DISABLED)
 const bool LayerManager::annotationsEnabled = false;
 #else
 const bool LayerManager::annotationsEnabled = true;
