diff --git a/src/mln/webgpu/context.cpp b/src/mln/webgpu/context.cpp
--- a/src/mln/webgpu/context.cpp
+++ b/src/mln/webgpu/context.cpp
@@ -59,6 +59,8 @@ Context::~Context() {
 
 void Context::beginFrame() {
     backend.getThreadPool().runRenderJobs();
+    stats.numDrawCalls = 0;
+    stats.numFrames++;
 }
 
 void Context::endFrame() {
diff --git a/src/mln/webgpu/drawable.cpp b/src/mln/webgpu/drawable.cpp
--- a/src/mln/webgpu/drawable.cpp
+++ b/src/mln/webgpu/drawable.cpp
@@ -975,6 +975,7 @@ void Drawable::draw(PaintParameters& parameters) const {
                                              baseInstance);                                // firstInstance
 
             context.renderingStats().numDrawCalls++;
+            context.renderingStats().totalDrawCalls++;
         }
     }
 }
