diff --git a/include/mln/i18n/number_format.hpp b/include/mln/i18n/number_format.hpp
index 6d1c6aeae1b0..5d75c91d13ca 100644
--- a/include/mln/i18n/number_format.hpp
+++ b/include/mln/i18n/number_format.hpp
@@ -1,6 +1,7 @@
 #pragma once
 
 #include <cstdint>
+#include <optional>
 #include <string>
 
 namespace mln {
@@ -9,8 +10,8 @@ namespace platform {
 std::string formatNumber(double number,
                          const std::string& localeId,
                          const std::string& currency,
-                         uint8_t minFractionDigits,
-                         uint8_t maxFractionDigits);
+                         std::optional<uint8_t> minFractionDigits,
+                         std::optional<uint8_t> maxFractionDigits);
 
 } // namespace platform
 } // namespace mln
diff --git a/platform/android/MapLibreAndroid/src/cpp/i18n/number_format.cpp b/platform/android/MapLibreAndroid/src/cpp/i18n/number_format.cpp
index c87e20455ba4..2427e64e1a8f 100644
--- a/platform/android/MapLibreAndroid/src/cpp/i18n/number_format.cpp
+++ b/platform/android/MapLibreAndroid/src/cpp/i18n/number_format.cpp
@@ -54,8 +54,8 @@ namespace platform {
 std::string formatNumber(double number,
                          const std::string& localeId,
                          const std::string& currency,
-                         uint8_t minFractionDigits,
-                         uint8_t maxFractionDigits) {
+                         std::optional<uint8_t> minFractionDigits,
+                         std::optional<uint8_t> maxFractionDigits) {
     auto env{android::AttachEnv()};
 
     jni::Global<jni::Object<android::Locale>> locale;
@@ -74,8 +74,10 @@ std::string formatNumber(double number,
     jni::Global<jni::Object<android::NumberFormat>> formatter;
     if (currency.empty()) {
         formatter = jni::NewGlobal(*env, android::NumberFormat::getInstance(*env, locale));
-        android::NumberFormat::setMinimumFractionDigits(*env, formatter, static_cast<jni::jint>(minFractionDigits));
-        android::NumberFormat::setMaximumFractionDigits(*env, formatter, static_cast<jni::jint>(maxFractionDigits));
+        android::NumberFormat::setMinimumFractionDigits(
+            *env, formatter, static_cast<jni::jint>(minFractionDigits.value_or(0)));
+        android::NumberFormat::setMaximumFractionDigits(
+            *env, formatter, static_cast<jni::jint>(maxFractionDigits.value_or(3)));
     } else {
         formatter = jni::NewGlobal(*env, android::NumberFormat::getCurrencyInstance(*env, locale));
     }
diff --git a/platform/android/src/test/number_format_test_stub.cpp b/platform/android/src/test/number_format_test_stub.cpp
index ca38b354123b..eaa554464cd7 100644
--- a/platform/android/src/test/number_format_test_stub.cpp
+++ b/platform/android/src/test/number_format_test_stub.cpp
@@ -6,8 +6,8 @@ namespace platform {
 std::string formatNumber(double /*number*/,
                          const std::string& /*localeId */,
                          const std::string& /*currency*/,
-                         uint8_t /*minFractionDigits*/,
-                         uint8_t /*maxFractionDigits*/
+                         std::optional<uint8_t> /*minFractionDigits*/,
+                         std::optional<uint8_t> /*maxFractionDigits*/
 ) {
     return "";
 }
diff --git a/platform/darwin/core/number_format.mm b/platform/darwin/core/number_format.mm
index 875ef7e48a98..95f632514313 100644
--- a/platform/darwin/core/number_format.mm
+++ b/platform/darwin/core/number_format.mm
@@ -6,7 +6,8 @@
 namespace platform {
 
 std::string formatNumber(double number, const std::string& localeId, const std::string& currency,
-                         uint8_t minFractionDigits, uint8_t maxFractionDigits) {
+                         std::optional<uint8_t> minFractionDigits,
+                         std::optional<uint8_t> maxFractionDigits) {
   // Create a new formatter each time to avoid state pollution between calls
   NSNumberFormatter* numberFormatter = [[NSNumberFormatter alloc] init];
 
@@ -15,8 +16,8 @@
   numberFormatter.currencyCode = !currency.empty() ? @(currency.c_str()) : nil;
 
   if (currency.empty()) {
-    numberFormatter.minimumFractionDigits = minFractionDigits;
-    numberFormatter.maximumFractionDigits = maxFractionDigits;
+    numberFormatter.minimumFractionDigits = minFractionDigits.value_or(0);
+    numberFormatter.maximumFractionDigits = maxFractionDigits.value_or(3);
     numberFormatter.numberStyle = NSNumberFormatterDecimalStyle;
   } else {
     numberFormatter.numberStyle = NSNumberFormatterCurrencyStyle;
diff --git a/platform/default/src/mln/i18n/number_format.cpp b/platform/default/src/mln/i18n/number_format.cpp
index 131595a7cb57..9ea6e77663cf 100644
--- a/platform/default/src/mln/i18n/number_format.cpp
+++ b/platform/default/src/mln/i18n/number_format.cpp
@@ -14,8 +14,8 @@ namespace platform {
 std::string formatNumber(double number,
                          const std::string& localeId,
                          const std::string& currency,
-                         uint8_t minFractionDigits,
-                         uint8_t maxFractionDigits) {
+                         std::optional<uint8_t> minFractionDigits,
+                         std::optional<uint8_t> maxFractionDigits) {
     UErrorCode status = U_ZERO_ERROR;
     icu::UnicodeString ustr;
     std::string formatted;
@@ -31,7 +31,8 @@ std::string formatNumber(double number,
                    .toString(status);
     } else {
         ustr = icu::number::NumberFormatter::with()
-                   .precision(icu::number::Precision::minMaxFraction(minFractionDigits, maxFractionDigits))
+                   .precision(icu::number::Precision::minMaxFraction(minFractionDigits.value_or(0),
+                                                                     maxFractionDigits.value_or(3)))
                    .locale(locale)
                    .formatDouble(number, status)
                    .toString(status);
@@ -39,7 +40,8 @@ std::string formatNumber(double number,
     return ustr.toUTF8String(formatted);
 }
 #else
-std::string formatNumber(double number, const std::string&, const std::string&, uint8_t, uint8_t) {
+std::string formatNumber(
+    double number, const std::string&, const std::string&, std::optional<uint8_t>, std::optional<uint8_t>) {
     return std::to_string(number);
 }
 #endif
diff --git a/platform/qt/src/mln/number_format.cpp b/platform/qt/src/mln/number_format.cpp
index e64dfa171d15..b049708c02e3 100644
--- a/platform/qt/src/mln/number_format.cpp
+++ b/platform/qt/src/mln/number_format.cpp
@@ -1,4 +1,4 @@
-#include <mln/util/platform.hpp>
+#include <mln/i18n/number_format.hpp>
 
 #include <QLocale>
 #include <QString>
@@ -9,8 +9,8 @@ namespace platform {
 std::string formatNumber(double number,
                          const std::string& localeId,
                          const std::string& currency,
-                         uint8_t minFractionDigits,
-                         uint8_t maxFractionDigits) {
+                         std::optional<uint8_t> minFractionDigits,
+                         std::optional<uint8_t> maxFractionDigits) {
     QString formatted;
     // Qt Locale::toString() API takes only one precision argument
     (void)minFractionDigits;
@@ -19,7 +19,7 @@ std::string formatNumber(double number,
     if (!currency.empty()) {
         formatted = locale.toCurrencyString(number);
     } else {
-        formatted = locale.toString(number, 'f', maxFractionDigits);
+        formatted = locale.toString(number, 'f', maxFractionDigits.value_or(3));
     }
     return formatted.toStdString();
 }
diff --git a/src/mln/style/expression/collator_expression.cpp b/src/mln/style/expression/collator_expression.cpp
index e60723b4f2d8..53288b6f78d1 100644
--- a/src/mln/style/expression/collator_expression.cpp
+++ b/src/mln/style/expression/collator_expression.cpp
@@ -1,3 +1,5 @@
+#include <exception>
+
 #include <mln/style/expression/collator.hpp>
 #include <mln/style/expression/collator_expression.hpp>
 #include <mln/style/expression/literal.hpp>
@@ -103,7 +105,7 @@ mln::Value CollatorExpression::serialize() const {
     return std::vector<mln::Value>{{std::string("collator"), options}};
 }
 
-EvaluationResult CollatorExpression::evaluate(const EvaluationContext& params) const {
+EvaluationResult CollatorExpression::evaluate(const EvaluationContext& params) const try {
     auto caseSensitiveResult = caseSensitive->evaluate(params);
     if (!caseSensitiveResult) {
         return caseSensitiveResult.error();
@@ -124,6 +126,8 @@ EvaluationResult CollatorExpression::evaluate(const EvaluationContext& params) c
     } else {
         return Collator(caseSensitiveResult->get<bool>(), diacriticSensitiveResult->get<bool>());
     }
+} catch (const std::exception& error) {
+    return EvaluationError{error.what()};
 }
 
 } // namespace expression
diff --git a/src/mln/style/expression/comparison.cpp b/src/mln/style/expression/comparison.cpp
index 2e04376dce82..4fb98c585935 100644
--- a/src/mln/style/expression/comparison.cpp
+++ b/src/mln/style/expression/comparison.cpp
@@ -1,3 +1,5 @@
+#include <exception>
+
 #include <mln/style/expression/collator.hpp>
 #include <mln/style/expression/comparison.hpp>
 #include <mln/style/expression/dsl.hpp>
@@ -56,22 +58,22 @@ bool gteq(const Value& lhs, const Value& rhs) noexcept {
                      });
 }
 
-bool eqCollate(const std::string& a, const std::string& b, const Collator& c) noexcept {
+bool eqCollate(const std::string& a, const std::string& b, const Collator& c) {
     return c.compare(a, b) == 0;
 }
-bool neqCollate(const std::string& a, const std::string& b, const Collator& c) noexcept {
+bool neqCollate(const std::string& a, const std::string& b, const Collator& c) {
     return !eqCollate(a, b, c);
 }
-bool ltCollate(const std::string& a, const std::string& b, const Collator& c) noexcept {
+bool ltCollate(const std::string& a, const std::string& b, const Collator& c) {
     return c.compare(a, b) < 0;
 }
-bool gtCollate(const std::string& a, const std::string& b, const Collator& c) noexcept {
+bool gtCollate(const std::string& a, const std::string& b, const Collator& c) {
     return c.compare(a, b) > 0;
 }
-bool lteqCollate(const std::string& a, const std::string& b, const Collator& c) noexcept {
+bool lteqCollate(const std::string& a, const std::string& b, const Collator& c) {
     return c.compare(a, b) <= 0;
 }
-bool gteqCollate(const std::string& a, const std::string& b, const Collator& c) noexcept {
+bool gteqCollate(const std::string& a, const std::string& b, const Collator& c) {
     return c.compare(a, b) >= 0;
 }
 
@@ -178,7 +180,7 @@ CollatorComparison::CollatorComparison(std::string op_,
                             (lhs->getType() == type::Value || rhs->getType() == type::Value);
 }
 
-EvaluationResult CollatorComparison::evaluate(const EvaluationContext& params) const {
+EvaluationResult CollatorComparison::evaluate(const EvaluationContext& params) const try {
     EvaluationResult lhsResult = lhs->evaluate(params);
     if (!lhsResult) return lhsResult;
 
@@ -196,6 +198,8 @@ EvaluationResult CollatorComparison::evaluate(const EvaluationContext& params) c
 
     const Collator& c = collatorResult->get<Collator>();
     return compare(lhsResult->get<std::string>(), rhsResult->get<std::string>(), c);
+} catch (const std::exception& error) {
+    return EvaluationError{error.what()};
 }
 
 void CollatorComparison::eachChild(const std::function<void(const Expression&)>& visit) const {
diff --git a/src/mln/style/expression/number_format.cpp b/src/mln/style/expression/number_format.cpp
index d6c6ce21a66e..8e7f5bafccba 100644
--- a/src/mln/style/expression/number_format.cpp
+++ b/src/mln/style/expression/number_format.cpp
@@ -1,3 +1,5 @@
+#include <exception>
+
 #include <mln/i18n/number_format.hpp>
 #include <mln/style/conversion_impl.hpp>
 #include <mln/style/expression/number_format.hpp>
@@ -28,7 +30,7 @@ NumberFormat::NumberFormat(std::unique_ptr<Expression> number_,
 
 NumberFormat::~NumberFormat() = default;
 
-EvaluationResult NumberFormat::evaluate(const EvaluationContext& params) const {
+EvaluationResult NumberFormat::evaluate(const EvaluationContext& params) const try {
     auto numberResult = number->evaluate(params);
     if (!numberResult) {
         return numberResult.error();
@@ -53,7 +55,7 @@ EvaluationResult NumberFormat::evaluate(const EvaluationContext& params) const {
         evaluatedCurrency = toString(*currencyResult);
     }
 
-    uint8_t evaluatedMinFractionDigits = 0;
+    std::optional<uint8_t> evaluatedMinFractionDigits;
     if (minFractionDigits) {
         auto minDigitsResult = minFractionDigits->evaluate(params);
         if (!minDigitsResult) {
@@ -62,7 +64,7 @@ EvaluationResult NumberFormat::evaluate(const EvaluationContext& params) const {
         evaluatedMinFractionDigits = static_cast<uint8_t>(minDigitsResult->get<double>());
     }
 
-    uint8_t evaluatedMaxFractionDigits = 3;
+    std::optional<uint8_t> evaluatedMaxFractionDigits;
     if (maxFractionDigits) {
         auto maxDigitsResult = maxFractionDigits->evaluate(params);
         if (!maxDigitsResult) {
@@ -75,6 +77,8 @@ EvaluationResult NumberFormat::evaluate(const EvaluationContext& params) const {
     output = platform::formatNumber(
         evaluatedNumber, evaluatedLocale, evaluatedCurrency, evaluatedMinFractionDigits, evaluatedMaxFractionDigits);
     return output;
+} catch (const std::exception& error) {
+    return EvaluationError{error.what()};
 }
 
 void NumberFormat::eachChild(const std::function<void(const Expression&)>& visit) const {
